packages/langium/src/grammar/validation/types-validator.ts

```diff
@@ -233,120 +227,63 @@ function validateDeclaredAndInferredConsistency(typeInfo: InferredInfo & Declare
-                const alias_branches: string[][] = JSON.parse(JSON.stringify(backup_branches));
-                alias_branches.forEach(alias_branch => alias_branch.push(alias));
-                branches.push(...alias_branches);
-            }
-        }
-    }
-    return branches.map(branch => distinctAndSorted(branch).join(' | '));
-}
-
-function typeAsStringKeywordsReplacement(found: PropertyType): string {
-    const propTypeWithStr = found.types.filter(e => !e.startsWith('\''));
-    propTypeWithStr.push('string');
-    return distinctAndSorted(propTypeWithStr).join(' | ');
-}
-
-function checkAlternativesConsistencyHelper(found: PropertyType[], expected: PropertyType[], typeToAliases: Map<string, Set<string>>): ErrorInfo[] {
-    const arrayReferenceError = (found: PropertyType, expected: PropertyType) =>
-        found.array && !expected.array && found.reference && !expected.reference ? 'can\'t be an array and a reference' :
-            !found.array && expected.array && !found.reference && expected.reference ? 'has to be an array and a reference' :
-                found.array && !expected.array ? 'can\'t be an array' :
-                    !found.array && expected.array ? 'has to be an array' :
-                        found.reference && !expected.reference ? 'can\'t be a reference' :
-                            !found.reference && expected.reference ? 'has to be a reference' : '';
-
-    const stringToFound = found.reduce((acc, propType) => acc.set(distinctAndSorted(propType.types).join(' | '), propType),
-        new Map<string, PropertyType>());
-
-    const stringToExpected = expected.reduce((acc, propType) => {
-        getAllAliases(propType, typeToAliases).forEach(alias => acc.set(alias, propType));
-        return acc;
-    }, new Map<string, PropertyType>());
-    const errorsInfo: ErrorInfo[] = [];
-
-    // detects extra type alternatives & check matched ones on consistency by 'array' and 'reference'
-    for (const [typeAsString, foundPropertyType] of stringToFound) {
-        const expectedPropertyType = stringToExpected.get(typeAsString) ?? stringToExpected.get(typeAsStringKeywordsReplacement(foundPropertyType));
-        if (!expectedPropertyType) {
-            errorsInfo.push({ typeAsString, errorMessage: 'is not expected' });
-        } else if (expectedPropertyType.array !== foundPropertyType.array || expectedPropertyType.reference !== foundPropertyType.reference) {
-            errorsInfo.push({ typeAsString, errorMessage: arrayReferenceError(foundPropertyType, expectedPropertyType) });
-        }
+function validateAlternativesConsistency(
+    inferred: PropertyType,
+    declared: PropertyType,
+    applyErrorToInferredTypes: (errorMessage: string) => void
+) {
+    if (!isTypeAssignable(inferred, declared)) {
+        applyErrorToInferredTypes(`Cannot assign type '${propertyTypeToString(inferred, 'DeclaredType')}' to '${propertyTypeToString(declared, 'DeclaredType')}'`);
```

Comment:
Shouldn't the first one be 'InferredType' instead of 'DeclaredType'?

Reply:
This is only for display. The other option would be `AstNode` which is for generation purposes only.